### ISSUE TITLE
[Discussion] .NET Core 2.0 Preview 2 is being rolled out to App Service

### ISSUE THREAD

**Issue by davidebbo** [Azure/app-service-announcements-discussions#2]
Discussion for https://github.com/Azure/app-service-announcements/issues/14

**Comment by ssougnez** [comment]
Hi, can I gently ask if there are any news on this? I'm developing an asp.net core 2.0 web application since 1 or 2 weeks and I'd like to be able to test it on Azure.

**Comment by suwatch** [comment]
@ssougnez It will be available on Azure by the end of the week (at the latest, next Mon).

**Comment by ssougnez** [comment]
All right, thanks ;-)

**Comment by johannordin** [comment]
@ssougnez you can still deploy 2.0.0-preview1-final if that is good enough for you :)

**Comment by ssougnez** [comment]
Well it was just to know if the roll out was going to happen soon or in 2 months ^^

Now that I know that it will be done in upcoming days, I can wait, but thanks for the advice ;-)

**Comment by suwatch** [comment]
@ssougnez We have completely rollout .NET Core 2.0 Preview 2.   Please give a try.

**Comment by ssougnez** [comment]
I still have the issue (http://dev-areaprog.azurewebsites.net/), I'll have a look at the log when I can to see if it's still the same error.

**Comment by watashiSHUN** [comment]
@ssougnez investigating it now

**Comment by ssougnez** [comment]
I just tried a 

`dotnet AreaProg.dll`

and get:

> Unhandled Exception: System.IO.FileLoadException: Could not load file or assembly 'Microsoft.AspNetCore.Hosting.Abstractions, Version=2.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60'. The located assembly's manifest definition does not match the assembly reference. (Exception from HRESULT: 0x80131040)
>    at AreaProg.Program.Main(String[] args)


Looks like it has to do with the DLL version which is weird. Indeed, The project has been created with Visual Studio 15.3 and contains this:

`<PackageReference Include="Microsoft.AspNetCore.All" Version="2.0.0-preview2-final" />`

So, it seems to be the correct version...

**Comment by watashiSHUN** [comment]
@ssougnez  could you give me more information on how you create your project in VS? I looked at your binaries, `Microsoft.AspNetCore.Hosting.Abstractions.dll` has version `1.1.2` (actually a lot of the dlls has this version), I created a simple dotnet 2.0 project myself and mine is `2.0.0-preview2-25794` which is also the latest on [nuget](https://www.nuget.org/packages/Microsoft.AspNetCore.Hosting.Abstractions/2.0.0-preview2-final)

based on your `.csproj` snippet, it looks like you want to create a 2.0 project but your assemblies are of 1.1
did you get any error when doing the `nuget restore`? can you run your website locally? do you have the 2.0 preview2 sdk and runtime installed locally?

**Comment by ssougnez** [comment]
Héhé, that's indeed all the questions I asked myself :-)

Before answering them, let me mention that the web site I gave you has some development in it, but I also tried on another web site (http://dev-robot-war.azurewebsites.net/) and this one contains a fresh new project created by Visual Studio 15.3 few days ago, without any modifications.

However, maybe the answer is in the version of .net core I have installed. When I run `dotnet --version`, I get 2.0.0-preview2-006497 (although, I found this version on dot.net, so I though it was the good one).

About how I created the Visual Studio project, I did it using the GUI, so "File > New Project >ASP.NET Core Web Application (.NET Core)", then I ensure that "netcore 2.0" was selected, then I validated. 

Concerning the nuget restore, I didn't pay a close attention to it, I just waited for it to be finished and everything was OK.

I'm actually currently testing my site on local and it works like a charm (note that both .net core 1.1 and .net core 2.0 are installed).

And as for last question, yes, the sdk is installed locally, but as stated at the beginning of the post, it seems not to be the same version than yours, however, I don't know if that's the issue.

Concerning the version of the DLL, I also noticed it few days ago on a web app that was at first based on asp.net core 1.1 and that I migrated, that's why I decided to create a brand new asp.net core 2.0 web app that I deployed in here "dev-robot-war.azurewebsites.net", however, this did not change anything.

As a complementary information, I checked the repositories of NuGet used by Visual Studio and I currently have :

```
https://api.nuget.org/v3/index.json
C:\Users\ssougnez\.dotnet\NuGetFallbackFolder
```

Is that the correct repositories ?

Thanks for your help ;-)

**Comment by watashiSHUN** [comment]
since you can run it locally, here are my two guesses:

1. it is never truly running 2.0 but instead, it's always 1.0

-  to test this theory, can you `rebuild solution` in VS, wait for `nuget restore` to finish
-  navigate to nuget cache (if you are running windows, try this command `cd %USERProfile%\.nuget\packages\Microsoft.AspNetCore.Hosting.Abstractions\`)
-  check if you have a folder `2.0.0-preview2-final` and verify its modify date

2. during debugging, it is using 2.0 assemblies, but during publish 1.0 assemblies got mixed in

- in VS choose `publish the project to folder`
- go to the publish folder and run `dotnet yourdll` to see if it was successful
- go to the `bin\Debug\' and run the same command to compare

**Comment by ssougnez** [comment]
so, for you guesses:

1. I do have a "2.0.0-preview2-final" folder and after the build, its modify date is 28.06.2017, however, even though it's not today, it's the most recently modified.

2. I did what you said and both result are the same. The application starts but I get an error, however, this is related to Entity Framework and empty connection string (as I couldn't start the app on Azure, I didn't work on the multi environment variable yet), so I think that at the end of the day, the app is able to launch.

I'd like to check the version of the "Microsoft.AspNetCore.Hosting.Abstractions" DLL but I have no idea where to find it. Is it the version of the installed SDK that is taken into account ?

Should i uninstall .netcore 1.1 to see if it solves the issue? (or create it on my local machine :p)

**Comment by Ponant** [comment]
@watashiSHUN , I am getting this as an error on the browser....
```
An error occurred while starting the application.

.NET Core 4.6.00001.0 X86 v4.0.0.0    |   Microsoft.AspNetCore.Hosting version 2.0.0-preview2-25794    |    Microsoft Windows 6.2.9200    |   Need help?
```

**Comment by watashiSHUN** [comment]
@ssougnez , to check the version, simply right click the dll, then choose `properties`
you will see a tab called `Details` it will display product version:

![capture](https://user-images.githubusercontent.com/6531400/27976630-3b36df1e-631c-11e7-9fd4-5691ee3dcad5.JPG)

**Comment by ssougnez** [comment]
Sorry, I was not clear enough :p I meant that I don't know how to find the DLL (not its version). I checked in the path you gave me (%USERProfile%\.nuget\packages\Microsoft.AspNetCore.Hosting.Abstractions\) and in there, the version corresponds to the folder name, so I think it's good.

At the end of the day, what I really meant is "where does file that gets uploaded to my Azure web site is?". Indeed, there is a "Microsoft.AspNetCore.Hosting.Abstractions" in my "D:\home\site\wwwroot" Azure directory, where does it come from? When I publish my web app in a folder, this DLL does not end up in the destination folder, so where is it taken from? Is it the one of the SDK? There must be a reason why its version is not the correct one.

**Comment by watashiSHUN** [comment]
are you saying that in your publish folder, you can't find `Microsoft.AspNetCore.Hosting.Abstractions.dll`?
this is weird, whatever end up in your publish folder will end up in the `D:\home\site\wwwroot` when you publish to Azure

**Comment by ssougnez** [comment]
That's exactly what I'm saying. When I run

` Get-ChildItem "*.*" -Recurse | ? { $_.Name.Contains("AspNetCore") }`

In the local folder where I published my web app, I get:

>     Directory: C:\Users\ssougnez\Desktop\az
> 
> 
> Mode                LastWriteTime         Length Name
> ----                -------------         ------ ----
> -a----       19/06/2017     20:29          55304 Microsoft.AspNetCore.Mvc.Razor.ViewCompilation.dll

Which is why I'm really wondering where the DLL on Azure comes from. Maybe an old publish that won't get away ? Is there a way to remove everything from the "D:\home\site\wwwroot" directory to try a fresh publish?

Note that I just went on "https://dev-areaprog.scm.azurewebsites.net/DebugConsole/" and I noticed that the file we're talking about has been last modified at "27/04/2017". Maybe I have an idea of what's happening.

As I'm currently learning Azure, I often created/deleted resources groups and app service, but I think that this one has been created few weeks ago and had, at some point in time, an asp.net core 1.1 web app installed on it. Then I recreated my project with asp.net core 2.0 (or I updated it, I really can't remember as I tried this with few project to get used to asp.net) and published it, but someone, I think that the DLL of the asp.net 1.1 project haven't been cleaned up correctly.

I tested something else that seems to confirm my assumption. I created a new asp.net core 1.1 web app with Visual Studio the same way I did the 2.0 one and tried to publish it in a folder and in that case, a lot of "Microsoft.AspNetCore.*" DLL are published in the folder. Then I try the exact same thing with an asp.net core 2.0 web app and the DLL were not published. 

In conclusion, I think that there is no issue with the roll out of .net core 2.0 on Azure (or at least, no issue related to this post) but the issue is that Visual Studio does not publish NuGet files along with the project. If you confirm this assumption, I'll report this bug in Visual Studio.

**Comment by watashiSHUN** [comment]
@Ponant can you give more detail? are you only seeing this in Azure? How did you deploy?

**Comment by watashiSHUN** [comment]
@ssougnez I think I might know how to fix your issue....
1. go to portal, stop your website
2. go to debug console, remove your wwwroot folder
3. in the portal, restart your site
4. do publish with VS again

the reason that the 2.0 does not have dlls is probably because dotnet core now has a `/store` folder (a new concept), which has all the assemblies required by `<PackageReference Include="Microsoft.AspNetCore.All" Version="2.0.0-preview2-final" />`

you can find it on your machine: `cd "%ProgramFiles(x86)%\dotnet\store"`

we also have this in Azure, thus, when you do `dotnet publish`, dlls will **not be included in your publish folder, but since you had an old project (1.1 which does not have \store...so assemblies are always included), the newly deployed 2.0 project will ACCIDENTALLY reference the 1.1 assemblies still remained in wwwroot**

once the old assemblies are removed from wwwroot, your project can automatically find 2.0 assemblies **already installed in Azure**

**Comment by watashiSHUN** [comment]
I guess this is a special case because the publish mechanism between 2 versions are now different...and usually we don't remove unnecessary files from wwwroot to speed up the publish steps

**Comment by ssougnez** [comment]
Hehe, look at that, it worked like a charm (although I had to publish it twice, apparently, Visual Studio does not like when no "wwwroot" are on the Azure server ^^).

There is an error right now but that's normal, I still have to update my code to work on Azure.

As you said, the issue comes from the two different publish mechanisms but you should expect this question to come back often. It might be interesting to include, in Visual Studio, an option to clean up the destination folder prior publishing. Anyway, I'll link this thread in my Stack Overflow question, maybe it'll help people.

Thanks a lot for your support ;-)

**Comment by watashiSHUN** [comment]
To summarize, you need to pay special attention if you are 
1. using **Visual Studio** to publish project
2. your project is a dotnet core **2.0** project.
 
you need to clean up your repository before deployment: 
remove all the assemblies in your `D:\home\site\wwwroot` folder before publish to Azure

**Comment by ssougnez** [comment]
Indeed but I would add that these considerations should, I think, only be taken in the case of a migration from 1.x to 2.0. Otherwise, the 1.x DLL won't be there in the first place. The problem will solve by itself in few months but in the meanwhile, a lot of people might encounter it ;-)

**Comment by romanov** [comment]
What am i doing wrong?

1. I've downloaded and installed dotnet-sdk-2.0.0-preview2-006497
2. Created a new blank ASP.NET WEB CORE 2.0 website
3. Published to a folder and uploaded to my azure app server.

Got **HTTP Error 502.5 - Process Failure**

**Comment by watashiSHUN** [comment]
can you share your website URL either directly or [indirectly](https://github.com/projectkudu/kudu/wiki/Reporting-your-site-name-without-posting-it-publicly) so I can take a look

**Comment by romanov** [comment]
@watashiSHUN http://countryball.azurewebsites.net/

**Comment by watashiSHUN** [comment]
@romanov we just fix the VM for you, please verify

**Comment by romanov** [comment]
@watashiSHUN okay, the blank MVC is working. Can i test with it now with a SignalR core preview ?

**Comment by watashiSHUN** [comment]
@romanov yes, as long as you can run it locally, you should be able to do it in Azure

**Comment by Ponant** [comment]
@watashiSHUN , Sorry for the lack of detail and jet lag. Yes, I deployed to azure, via VS 2017 latest preview version with SDK preview 2. The app works locally, no issue here. Notice that the web app on azure was already run by a .Net core (1.0 or 1.1, I do not remember). Perhaps that will be solved if I deploy it on a fresh new slot on azure?

**Comment by watashiSHUN** [comment]
@Ponant I would expect that deploy to a fresh web app to work, but your error message looks different than ssougnez's, the fact that you **are not getting a 500 http status code means that dotnet core is indeed running**, if you can share your website name directly or [indirectly](https://github.com/projectkudu/kudu/wiki/Reporting-your-site-name-without-posting-it-publicly) I will take a look

**Comment by Ponant** [comment]
I am trying to send it on a new web app. I am on the West EU region.

**Comment by Ponant** [comment]
@watashiSHUN , it is not working for a fresh web app neither. Same error, here is the link
https://watashishun.azurewebsites.net/

**Comment by Ponant** [comment]
@watashiSHUN , ok It does not work when I use my project and uploaded to a new web app on azure, but it works if I deploy a ASP.NET core 2.0 preview 2 template from VS. So this seems to be due to my project. The problem is that I do not know how to resolve this because the error is vague. The project I have was created from an empty template Core 2.0 preview 2 (.Net 4.7) and I added three .Net standard 2.0 libraries that it references.

**Comment by watashiSHUN** [comment]
@Ponant 
if I understand correctly, you have 2 ways to deploy
1. use VS2017 publish to Azure (and you are saying that this is working?)
2. upload your project (how do you do this? how did you build? did you see any error during build? and how did you push it to Azure?)

you mentioned that your app does work locally, I downloaded the binary and running `dotnet yourdll` it gives me the same error as in Azure. I would think **your** local version is different than what's in https://watashishun.azurewebsites.net/?

```
Unhandled Exception: System.ArgumentException: The path in 'value' must start with '/'.
Parameter name: value
   at Microsoft.AspNetCore.Http.PathString..ctor(String value)
```

**Comment by Ponant** [comment]
@watashiSHUN , I am using VS 2017 for publishing in all cases.
I am investigating you error, but what I do not understand is why the dll would not be the same locally or on azure?

**Comment by watashiSHUN** [comment]
@Ponant you can look at the full error message in `D:\home\site\logs>` at https://watashishun.scm.azurewebsites.net.

In VS, if you try to publish to folder, and `cd` into that folder, then run `dotnet yourdll`, if this works then it means local site is running fine which I suspect will fail?

**Comment by Ponant** [comment]
@watashiSHUN , I fixed it thanks for your hint. I will now delete this temporary web app. 

The reason was that I am using attribute routing and one of my route was "lan" (for language url), so I turned it to "/lan" and now it works. What I do not understand is why "lan" worked locally on IIS express but not on azure, probably a web.config issue or the like?

**Comment by davidebbo** [comment]
I'd rather not use this thread as a catch all for all Core issues on App Service. It was only meant to discuss the status of the deployment, which is now complete.

I would suggest either StackOverflow questions or posts in the Web App MSDN forums to discuss individual issues you may run into. Thanks!

**Comment by robertmclaws** [comment]
So @Romanov's post about the STDOUT error had the same thing my app did, in that it said that **.NET Core 2.0 was installed using a Mac and Linux profile on Windows machines**. Could that be causing the problem?

Also, I **temporarily** solved this problem in my instance by doing a "self-contained" deployment, through adding `<RuntimeIdentifiers>win10-x64</RuntimeIdentifiers>` to my .csproj file. That still doesn't solve the core problem that I believe Core 2.0 Preview 2 is not installed correctly on these Web Apps VMs.

**Comment by watashiSHUN** [comment]
@robertmclaws can you provide a webapp name I will take a look, thanks

**Comment by robertmclaws** [comment]
@watashiSHUN It's "accrue-beta". Thanks!

**Comment by watashiSHUN** [comment]
```
Error:
  An assembly specified in the application dependencies manifest (Accrue.Web.deps.json) was not found:
    package: 'Microsoft.AspNetCore.Mvc.Razor.ViewCompilation', version: '2.0.0-preview2-final'
    path: 'lib/netcoreapp2.0/Microsoft.AspNetCore.Mvc.Razor.ViewCompilation.dll'

  This assembly was expected to be in the local runtime store as the application was published using the following target manifest files:
    aspnetcore-store-2.0.0-preview2-final-osx-x64.xml;aspnetcore-store-2.0.0-preview2-final-linux-x64.xml
```

is this the error message you are seeing before doing a self-contained deployment? I believe that `Microsoft.AspNetCore.Mvc.Razor.ViewCompilation.dll` is **generated during publish** (compiled view), it does not come with dotnet core 2.0.

so when I am debugging, after doing `dotnet publish`, this dll will be put into my output folder. although I see that you are not using this dll for your app, somehow it is still looking for it?
I probably need to know about your build process, but in short it does not seem to be a dotnet core runtime issue

**Comment by robertmclaws** [comment]
Yes, that was the error, and I can confirm that a) that file was NOT copied from the NuGet cache during the default publish, and b) there was no apparent way to tell VS to take THAT package and include it, since it is referenced by Microsoft.AspNetCore.All.

The solution I stated above triggered a "self-contained" publish, which took every single .NET core file and copied it to the output. That's why it's there now and wasn't there before.

Would it help if I published the default way to another Azure Web App so you can see what VS does?

Thanks!

**Comment by watashiSHUN** [comment]
@robertmclaws yes, that should help, what is the `dotnet core sdk` version you have installed locally? (`2.0.0-preview2-006497` this is what I use to publish, and I see the dll is being copied), even though it sounds like you are using VS (which is msbuild) I believe it still uses your local sdk

**Comment by watashiSHUN** [comment]
@ALL for any dotnet core 2.0 problems, please open an issue [here](https://github.com/projectkudu/kudu/issues/new) so we can better assist you

**Comment by Niyo** [comment]
I still have version 2.0.0-preview-1005977 on my app service in german cloud installed. Is german cloud handled seperatly with a different deployment date for preview2? If yes, where can I find more informtions?

**Comment by watashiSHUN** [comment]
@Niyo can you give me a webapp name, I will verify the installation on your machine?

**Comment by Niyo** [comment]
@watashiSHUN Okay thank you. No clear ETA? Is there a website or sth like this where I could get information about german cloud rollout dates the next time?

**Comment by xperiandri** [comment]
So had ASP.NET Core site extension only been required to run preview 1 until this rollout?

**Comment by martincostello** [comment]
I had a similar problem with this on an App Service that `2.0.0` had been rolled out to, and the root cause was that ASP.NET Core assemblies that 1.x published to `wwwroot` in a previous version were being loaded by the app instead of the `2.0.0` versions in the runtime package store. The Visual Studio publish profile wasn't set up to delete files already at the destination, causing the problem.

**Comment by holmsq** [comment]
German App Service is still using old version (2.0.0-preview2-006497) of Asp.Net core. When is going to be an update to 2.0

**Comment by davidebbo** [comment]
@holmsq ETA is by the end of August.

**Comment by jonasTMoreira** [comment]
Guys, I'm having the same problem, but I haven't been able to solve it yet.

Can you help?


http://mbcore.azurewebsites.net/

**Comment by davidebbo** [comment]
@jonasTMoreira this issue was tracking the release of a Preview of the 2.0 runtime almost a year ago, so this is very unlikely to be relevant to your scenario. I suggest opening a new StackOverflow question for your case (feel free to link from here).

**Comment by jonasTMoreira** [comment]
Okay Davidebbo. I'm dying to fix this problem but I couldn't.

**Comment by davidebbo** [comment]
@jonasTMoreira Happy to help. You just landed in completely the wrong place. :) One you have it on SO, I can help take a look.

**Comment by davidebbo** [comment]
@jonasTMoreira actually, I found your issue. Just remove `-argFile IISExeLauncherArgs.txt` from your web.config and you'll be rolling. https://github.com/dotnet/cli/issues/7741

**Comment by jonasTMoreira** [comment]
Thank you davidebbo! You helped me continue the migration of the project here. Thank you very much! Have a great day :)
